File: cloudflare-api/src/training.ts
import type { Context, Hono } from "hono";
import { ApiProblem, adminUser, currentUser, enumField, jsonObject, nowIso, numberField, parsePagination, publicPlayer, publicTeam, stringField } from "./helpers";
import { canOpenTeam, requireAimzTeam, scopedTeams } from "./team-access";
import type { AttendanceRow, AvailabilityRow, PlayerRow, TeamRow, TrainingRow, UserRow } from "./types";

type App = Hono<{ Bindings: Env }>;

function validInstant(value: unknown, field = "starts_at"): string {
  if (typeof value !== "string" || Number.isNaN(Date.parse(value))) throw new ApiProblem(422, "validation_error", "Check the highlighted fields.", [{ field, message: "Enter a valid date and time." }]);
  return new Date(value).toISOString();
}

async function trainingById(env: Env, id: string): Promise<TrainingRow> {
  const row = await env.DB.prepare("SELECT * FROM training_sessions WHERE id=?").bind(id).first<TrainingRow>();
  if (!row) throw new ApiProblem(404, "training_not_found", "Training session not found.");
  return row;
}

async function trainingTeam(env: Env, teamId: string): Promise<TeamRow> {
  const team = await env.DB.prepare("SELECT * FROM teams WHERE id=?").bind(teamId).first<TeamRow>();
  if (!team) throw new ApiProblem(404, "team_not_found", "Squad not found.");
  return team;
}

function publicTraining(row: TrainingRow, team: TeamRow): Record<string, unknown> {
  return { ...row, team: publicTeam(team) };
}

async function requireTrainingAccess(c: Context<{ Bindings: Env }>, row: TrainingRow, user?: UserRow): Promise<UserRow> {
  const actor = user ?? await currentUser(c);
  if (actor.role === "admin") return actor;
  if (!(await canOpenTeam(c.env, actor, row.team_id))) throw new ApiProblem(403, "team_access_denied", "You can only open your own squad's training sessions.");
  return actor;
}

/**
 * The register for one session: every player on the squad, with how they were
 * marked, or null for anyone not marked yet.
 *
 * The squad comes from the roster rather than from the marks, so a coach
 * opening a session for the first time is handed the whole squad to mark. There
 * is no attendance roster of its own.
 */
async function attendanceFor(env: Env, session: TrainingRow): Promise<Record<string, unknown>> {
  const [squad, marks] = await Promise.all([
    env.DB.prepare("SELECT * FROM players WHERE team_id=? AND is_active=1 ORDER BY name").bind(session.team_id).all<PlayerRow>(),
    env.DB.prepare("SELECT * FROM training_attendance WHERE training_session_id=?").bind(session.id).all<AttendanceRow>(),
  ]);
  const byPlayer = new Map(marks.results.map((row) => [row.player_id, row]));
  const items = squad.results.map((player) => ({
    player: publicPlayer(player),
    status: byPlayer.get(player.id)?.status ?? null,
    marked_at: byPlayer.get(player.id)?.updated_at ?? null,
  }));
  return {
    items,
    present: items.filter((row) => row.status === "present").length,
    absent: items.filter((row) => row.status === "absent").length,
    unmarked: items.filter((row) => row.status === null).length,
  };
}

export function registerTrainingRoutes(app: App): void {
  app.get("/api/v1/training-sessions", async (c) => {
    const url = new URL(c.req.url);
    const requested = url.searchParams.get("team_id");
    const { teamIds: allowedTeamIds } = await scopedTeams(c, requested);
    const { limit, offset } = parsePagination(url);
    const conditions: string[] = [];
    const values: unknown[] = [];
    // A parent sees every squad their children are on, so this is an IN list.
    if (allowedTeamIds) { conditions.push(`team_id IN (${allowedTeamIds.map(() => "?").join(",")})`); values.push(...allowedTeamIds); }
    for (const [parameter, operator] of [["from", ">="], ["to", "<="]] as const) {
      const raw = url.searchParams.get(parameter);
      if (raw) { conditions.push(`starts_at ${operator} ?`); values.push(validInstant(raw, parameter)); }
    }
    const where = conditions.length ? ` WHERE ${conditions.join(" AND ")}` : "";
    const [count, rows] = await Promise.all([
      c.env.DB.prepare(`SELECT COUNT(*) total FROM training_sessions${where}`).bind(...values).first<{ total: number }>(),
      c.env.DB.prepare(`SELECT * FROM training_sessions${where} ORDER BY starts_at LIMIT ? OFFSET ?`).bind(...values, limit, offset).all<TrainingRow>(),
    ]);
    const teamIds = [...new Set(rows.results.map((row) => row.team_id))];
    const teams = teamIds.length ? await c.env.DB.prepare(`SELECT * FROM teams WHERE id IN (${teamIds.map(() => "?").join(",")})`).bind(...teamIds).all<TeamRow>() : { results: [] as TeamRow[] };
    const byId = new Map(teams.results.map((team) => [team.id, team]));
    return c.json({ items: rows.results.map((row) => publicTraining(row, byId.get(row.team_id)!)), total: count?.total ?? 0, limit, offset });
  });

  app.get("/api/v1/training-sessions/:id", async (c) => {
    const row = await trainingById(c.env, c.req.param("id"));
    await requireTrainingAccess(c, row);
    return c.json(publicTraining(row, await trainingTeam(c.env, row.team_id)));
  });

  app.post("/api/v1/training-sessions", async (c) => {
    await adminUser(c);
    const body = await jsonObject(c);
    const teamId = stringField(body, "team_id", { min: 1, max: 36 })!;
    await requireAimzTeam(c.env, teamId);
    const venue = stringField(body, "venue", { min: 2, max: 200 })!;
    const notes = stringField(body, "notes", { optional: true, nullable: true, max: 2000 }) ?? null;
    const duration = numberField(body, "duration_minutes", { min: 15, max: 300, integer: true })!;
    if (!Array.isArray(body.occurrences) || body.occurrences.length < 1 || body.occurrences.length > 200) throw new ApiProblem(422, "validation_error", "Add between 1 and 200 training occurrences.", [{ field: "occurrences", message: "Add between 1 and 200 dates." }]);
    const occurrences = [...new Set(body.occurrences.map((value) => validInstant(value, "occurrences")))].sort();
    const seriesId = occurrences.length > 1 ? crypto.randomUUID() : null;
    const now = nowIso();
    const rows = occurrences.map<TrainingRow>((startsAt) => ({ id: crypto.randomUUID(), team_id: teamId, starts_at: startsAt, duration_minutes: duration, venue, notes, series_id: seriesId, created_at: now, updated_at: now }));
    await c.env.DB.batch(rows.map((row) => c.env.DB.prepare("INSERT INTO training_sessions (id, team_id, starts_at, duration_minutes, venue, notes, series_id, created_at, updated_at) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)").bind(row.id, row.team_id, row.starts_at, row.duration_minutes, row.venue, row.notes, row.series_id, now, now)));
    const team = await trainingTeam(c.env, teamId);
    return c.json(rows.map((row) => publicTraining(row, team)), 201);
  });

  app.patch("/api/v1/training-sessions/:id", async (c) => {
    await adminUser(c);
    const current = await trainingById(c.env, c.req.param("id"));
    const body = await jsonObject(c);
    const row: TrainingRow = {
      ...current,
      starts_at: body.starts_at === undefined ? current.starts_at : validInstant(body.starts_at),
      duration_minutes: numberField(body, "duration_minutes", { optional: true, min: 15, max: 300, integer: true }) ?? current.duration_minutes,
      venue: stringField(body, "venue", { optional: true, min: 2, max: 200 }) ?? current.venue,
      notes: body.notes === undefined ? current.notes : stringField(body, "notes", { nullable: true, max: 2000 }) ?? null,
      updated_at: nowIso(),
    };
    await c.env.DB.prepare("UPDATE training_sessions SET starts_at=?, duration_minutes=?, venue=?, notes=?, updated_at=? WHERE id=?").bind(row.starts_at, row.duration_minutes, row.venue, row.notes, row.updated_at, row.id).run();
    return c.json(publicTraining(row, await trainingTeam(c.env, row.team_id)));
  });

  app.delete("/api/v1/training-sessions/:id", async (c) => {
    await adminUser(c);
    const row = await trainingById(c.env, c.req.param("id"));
    const scope = new URL(c.req.url).searchParams.get("scope") ?? "one";
    if (scope !== "one" && scope !== "series") throw new ApiProblem(422, "validation_error", "Delete one session or its whole series.");
    if (scope === "series" && row.series_id) await c.env.DB.prepare("DELETE FROM training_sessions WHERE series_id=?").bind(row.series_id).run();
    else await c.env.DB.prepare("DELETE FROM training_sessions WHERE id=?").bind(row.id).run();
    return c.body(null, 204);
  });

  app.get("/api/v1/training-sessions/:id/availability", async (c) => {
    const session = await trainingById(c.env, c.req.param("id"));
    await requireTrainingAccess(c, session);
    const rows = await c.env.DB.prepare("SELECT * FROM training_availability WHERE training_session_id=? ORDER BY updated_at DESC").bind(session.id).all<AvailabilityRow>();
    const playerIds = rows.results.map((row) => row.player_id);
    const players = playerIds.length ? await c.env.DB.prepare(`SELECT * FROM players WHERE id IN (${playerIds.map(() => "?").join(",")})`).bind(...playerIds).all<PlayerRow>() : { results: [] as PlayerRow[] };
    const byId = new Map(players.results.map((player) => [player.id, player]));
    return c.json(rows.results.map((row) => ({ ...row, player: publicPlayer(byId.get(row.player_id) ?? null) })));
  });

  /**
   * The register for one session: every player on the squad, with how they were
   * marked, or null for anyone not marked yet.
   *
   * The squad is read from the roster rather than from the rows, so a coach
   * opening a session for the first time is handed the whole squad to mark
   * rather than an empty list. There is no separate attendance roster.
   */
  app.get("/api/v1/training-sessions/:id/attendance", async (c) => {
    const session = await trainingById(c.env, c.req.param("id"));
    await requireTrainingAccess(c, session);
    return c.json(await attendanceFor(c.env, session));
  });

  /**
   * Marks players present or absent. Only what is sent is touched, so a coach
   * correcting one name does not clear the rest of the register, and sending
   * null for somebody takes their mark away rather than guessing at it.
   */
  app.put("/api/v1/training-sessions/:id/attendance", async (c) => {
    await adminUser(c);
    const session = await trainingById(c.env, c.req.param("id"));
    const body = await jsonObject(c);
    if (!Array.isArray(body.entries) || body.entries.length > 200) throw new ApiProblem(422, "validation_error", "Send between 1 and 200 attendance marks.", [{ field: "entries", message: "Send up to 200 marks." }]);
    const entries = body.entries.map((raw) => {
      const entry = (raw ?? {}) as Record<string, unknown>;
      const playerId = stringField(entry, "player_id", { min: 1, max: 36 })!;
      const status = entry.status === null || entry.status === undefined ? null : enumField(entry, "status", ["present", "absent"] as const);
      return { playerId, status };
    });
    // Every name has to be on this squad, so a register cannot quietly collect
    // players who were never going to be at it.
    const ids = [...new Set(entries.map((entry) => entry.playerId))];
    if (ids.length) {
      const onSquad = await c.env.DB.prepare(`SELECT id FROM players WHERE team_id=? AND id IN (${ids.map(() => "?").join(",")})`).bind(session.team_id, ...ids).all<{ id: string }>();
      const known = new Set(onSquad.results.map((row) => row.id));
      const stranger = ids.find((id) => !known.has(id));
      if (stranger) throw new ApiProblem(422, "player_not_found", "Mark only players from this squad.");
    }
    const now = nowIso();
    if (entries.length) {
      await c.env.DB.batch(entries.map((entry) => entry.status === null
        ? c.env.DB.prepare("DELETE FROM training_attendance WHERE training_session_id=? AND player_id=?").bind(session.id, entry.playerId)
        : c.env.DB.prepare(`INSERT INTO training_attendance (training_session_id, player_id, status, created_at, updated_at) VALUES (?, ?, ?, ?, ?)
            ON CONFLICT(training_session_id, player_id) DO UPDATE SET status=excluded.status, updated_at=excluded.updated_at`).bind(session.id, entry.playerId, entry.status, now, now)));
    }
    // The whole register comes back rather than an acknowledgement, so the
    // screen that marked one player has the new tallies without asking again.
    return c.json(await attendanceFor(c.env, session));
  });

  app.put("/api/v1/training-sessions/:id/availability", async (c) => {
    const actor = await currentUser(c);
    const session = await trainingById(c.env, c.req.param("id"));
    await requireTrainingAccess(c, session, actor);
    const body = await jsonObject(c);
    const playerId = actor.role === "admin"
      ? stringField(body, "player_id", { min: 1, max: 36 })!
      : actor.player_id!;
    const player = await c.env.DB.prepare("SELECT * FROM players WHERE id=? AND team_id=?").bind(playerId, session.team_id).first<PlayerRow>();
    if (!player) throw new ApiProblem(422, "player_not_found", "Choose a player from this squad.");
    const status = enumField(body, "status", ["going", "not_going"] as const);
    const note = stringField(body, "note", { optional: true, nullable: true, max: 500 }) ?? null;
    const existing = await c.env.DB.prepare("SELECT * FROM training_availability WHERE training_session_id=? AND player_id=?").bind(session.id, playerId).first<AvailabilityRow>();
    const now = nowIso();
    const row: AvailabilityRow = { id: existing?.id ?? crypto.randomUUID(), training_session_id: session.id, player_id: playerId, status, note, created_at: existing?.created_at ?? now, updated_at: now };
    await c.env.DB.prepare(`INSERT INTO training_availability (id, training_session_id, player_id, status, note, created_at, updated_at) VALUES (?, ?, ?, ?, ?, ?, ?)
      ON CONFLICT(training_session_id, player_id) DO UPDATE SET status=excluded.status, note=excluded.note, updated_at=excluded.updated_at`).bind(row.id, row.training_session_id, row.player_id, row.status, row.note, row.created_at, row.updated_at).run();
    return c.json({ ...row, player: publicPlayer(player) });
  });
}

export { requireTrainingAccess, trainingById };

File: backend/app/api/v1/routes/training.py
from datetime import datetime

from fastapi import APIRouter, Response
from sqlalchemy import delete, func, select
from sqlalchemy.orm import selectinload

from app.api.deps import CurrentUser, SessionDep, TeamOperator
from app.core.errors import api_error
from app.db.models import (
    Player,
    TrainingAvailability,
    TrainingSession,
    User,
    UserRole,
    new_id,
)
from app.schemas import (
    AvailabilityInput,
    AvailabilityRead,
    Page,
    PlayerRead,
    TeamRead,
    TrainingCreate,
    TrainingSessionRead,
    TrainingUpdate,
)
from app.services.team_access import (
    can_open_team,
    require_team_operator,
    scoped_teams,
)

router = APIRouter()


def _page_args(limit: int, offset: int) -> tuple[int, int]:
    return min(max(limit, 1), 100), max(offset, 0)


def _serialize(row: TrainingSession) -> TrainingSessionRead:
    data = TrainingSessionRead.model_validate(row)
    data.team = TeamRead.model_validate(row.team) if row.team else None
    return data


async def _session_or_404(session: SessionDep, training_id: str) -> TrainingSession:
    row = await session.scalar(
        select(TrainingSession)
        .where(TrainingSession.id == training_id)
        .options(selectinload(TrainingSession.team))
    )
    if row is None:
        raise api_error(404, "training_not_found", "Training session not found.")
    return row


async def _require_access(session: SessionDep, user: User, row: TrainingSession) -> None:
    if user.role == UserRole.admin:
        return
    if not await can_open_team(session, user, row.team_id):
        raise api_error(
            403,
            "team_access_denied",
            "You can only open your own squad's training sessions.",
        )


@router.get("/training-sessions", response_model=Page[TrainingSessionRead])
async def list_training(
    current_user: CurrentUser,
    session: SessionDep,
    team_id: str | None = None,
    date_from: datetime | None = None,
    date_to: datetime | None = None,
    limit: int = 50,
    offset: int = 0,
) -> Page[TrainingSessionRead]:
    team_ids = await scoped_teams(session, current_user, team_id)
    limit, offset = _page_args(limit, offset)
    query = select(TrainingSession).options(selectinload(TrainingSession.team))
    count_query = select(func.count()).select_from(TrainingSession)
    conditions = [
        TrainingSession.team_id.in_(team_ids) if team_ids is not None else None,
        TrainingSession.starts_at >= date_from if date_from else None,
        TrainingSession.starts_at <= date_to if date_to else None,
    ]
    for condition in conditions:
        if condition is not None:
            query, count_query = query.where(condition), count_query.where(condition)
    total = await session.scalar(count_query) or 0
    rows = (
        await session.scalars(query.order_by(TrainingSession.starts_at).limit(limit).offset(offset))
    ).all()
    return Page(items=[_serialize(row) for row in rows], total=total, limit=limit, offset=offset)


@router.get("/training-sessions/{training_id}", response_model=TrainingSessionRead)
async def get_training(
    training_id: str, current_user: CurrentUser, session: SessionDep
) -> TrainingSessionRead:
    row = await _session_or_404(session, training_id)
    await _require_access(session, current_user, row)
    return _serialize(row)


@router.post("/training-sessions", response_model=list[TrainingSessionRead], status_code=201)
async def create_training(
    payload: TrainingCreate, actor: TeamOperator, session: SessionDep
) -> list[TrainingSessionRead]:
    await require_team_operator(session, actor, payload.team_id)
    # Dedupe and order the occurrences; a recurring block shares one series id.
    occurrences = sorted(set(payload.occurrences))
    series_id = new_id() if len(occurrences) > 1 else None
    rows = [
        TrainingSession(
            team_id=payload.team_id,
            starts_at=starts_at,
            duration_minutes=payload.duration_minutes,
            venue=payload.venue,
            notes=payload.notes,
            series_id=series_id,
        )
        for starts_at in occurrences
    ]
    session.add_all(rows)
    await session.commit()
    ids = [row.id for row in rows]
    reloaded = (
        await session.scalars(
            select(TrainingSession)
            .where(TrainingSession.id.in_(ids))
            .options(selectinload(TrainingSession.team))
            .order_by(TrainingSession.starts_at)
        )
    ).all()
    return [_serialize(row) for row in reloaded]


@router.patch("/training-sessions/{training_id}", response_model=TrainingSessionRead)
async def update_training(
    training_id: str, payload: TrainingUpdate, actor: TeamOperator, session: SessionDep
) -> TrainingSessionRead:
    row = await _session_or_404(session, training_id)
    await require_team_operator(session, actor, row.team_id)
    provided = payload.model_fields_set
    if "starts_at" in provided and payload.starts_at is not None:
        row.starts_at = payload.starts_at
    if "duration_minutes" in provided and payload.duration_minutes is not None:
        row.duration_minutes = payload.duration_minutes
    if "venue" in provided and payload.venue is not None:
        row.venue = payload.venue
    if "notes" in provided:
        row.notes = payload.notes
    await session.commit()
    return _serialize(await _session_or_404(session, training_id))


@router.delete("/training-sessions/{training_id}", status_code=204)
async def delete_training(
    training_id: str,
    actor: TeamOperator,
    session: SessionDep,
    scope: str = "one",
) -> Response:
    if scope not in {"one", "series"}:
        raise api_error(422, "validation_error", "Delete one session or its whole series.")
    row = await _session_or_404(session, training_id)
    await require_team_operator(session, actor, row.team_id)
    if scope == "series" and row.series_id:
        await session.execute(
            delete(TrainingSession).where(TrainingSession.series_id == row.series_id)
        )
    else:
        await session.delete(row)
    await session.commit()
    return Response(status_code=204)


def _availability(row: TrainingAvailability, player: Player | None) -> AvailabilityRead:
    # Built field-by-field rather than model_validate so we never touch the
    # ``player`` relationship on ``row`` (which would trigger an async lazy load).
    return AvailabilityRead(
        id=row.id,
        training_session_id=row.training_session_id,
        player_id=row.player_id,
        status=row.status,
        note=row.note,
        created_at=row.created_at,
        updated_at=row.updated_at,
        player=PlayerRead.model_validate(player) if player else None,
    )


@router.get(
    "/training-sessions/{training_id}/availability",
    response_model=list[AvailabilityRead],
)
async def list_availability(
    training_id: str, current_user: CurrentUser, session: SessionDep
) -> list[AvailabilityRead]:
    row = await _session_or_404(session, training_id)
    await _require_access(session, current_user, row)
    entries = (
        await session.scalars(
            select(TrainingAvailability)
            .where(TrainingAvailability.training_session_id == training_id)
            .options(selectinload(TrainingAvailability.player))
            .order_by(TrainingAvailability.updated_at.desc())
        )
    ).all()
    return [_availability(entry, entry.player) for entry in entries]


@router.put(
    "/training-sessions/{training_id}/availability",
    response_model=AvailabilityRead,
)
async def set_availability(
    training_id: str,
    payload: AvailabilityInput,
    current_user: CurrentUser,
    session: SessionDep,
) -> AvailabilityRead:
    row = await _session_or_404(session, training_id)
    await _require_access(session, current_user, row)
    # An admin names the player; anyone else answers for their own linked player.
    if current_user.role == UserRole.admin:
        if not payload.player_id:
            raise api_error(422, "player_not_found", "Choose a player from this squad.")
        player_id = payload.player_id
    else:
        if current_user.player_id is None:
            raise api_error(422, "player_not_found", "Choose a player from this squad.")
        player_id = current_user.player_id
    player = await session.scalar(
        select(Player).where(Player.id == player_id, Player.team_id == row.team_id)
    )
    if player is None:
        raise api_error(422, "player_not_found", "Choose a player from this squad.")
    entry = await session.scalar(
        select(TrainingAvailability).where(
            TrainingAvailability.training_session_id == training_id,
            TrainingAvailability.player_id == player_id,
        )
    )
    if entry is None:
        entry = TrainingAvailability(training_session_id=training_id, player_id=player_id)
        session.add(entry)
    entry.status = payload.status
    entry.note = payload.note
    await session.commit()
    # Repopulate server-generated timestamps on a freshly-inserted row.
    await session.refresh(entry)
    return _availability(entry, player)
